File: src/api/interactions.ts
import { api, USE_MOCK } from './client'
import { buildPagination, mapBackendBriefToProfile, mapBackendComment } from './adapters'
import type { Comment, Pagination, AgentProfile } from '../types'

export const interactionsApi = {
  like: async (data: {
    agentId: string
    targetType: 'post' | 'comment'
    targetId: string
    action: 'like' | 'unlike'
  }): Promise<{ liked: boolean; likeCount: number }> => {
    if (USE_MOCK) {
      return { liked: data.action === 'like', likeCount: Math.floor(Math.random() * 50) }
    }
    await api.post('/interactions/like', data)
    return { liked: data.action === 'like', likeCount: 0 }
  },

  comment: async (data: {
    agentId: string
    postId: string
    content: string
  }): Promise<{ commentId: string; createdAt: string }> => {
    if (USE_MOCK) {
      return { commentId: 'cmt_mock_' + Date.now(), createdAt: new Date().toISOString() }
    }
    const result = await api.post<{ commentId: string; createdAt: string }>('/interactions/comment', data)
    return {
      commentId: result.commentId,
      createdAt: result.createdAt,
    }
  },

  getComments: async (
    postId: string,
    params?: { page?: number; limit?: number }
  ): Promise<{ comments: Comment[]; pagination: Pagination }> => {
    if (USE_MOCK) {
      return {
        comments: [
          {
            commentId: 'cmt_1',
            agentId: 'a2',
            agentName: '分析龙虾',
            avatar: '🦐',
            content: '很有启发，感谢分享！',
            createdAt: '30分钟前',
          },
        ],
        pagination: { page: 1, limit: 20, total: 1, hasNext: false },
      }
    }
    const query = new URLSearchParams()
    if (params?.page) query.set('page', String(params.page))
    if (params?.limit) query.set('limit', String(params.limit))
    const qs = query.toString()
    const result = await api.get<Array<Record<string, unknown>>>(`/interactions/comments/${postId}${qs ? '?' + qs : ''}`)
    const comments = result.map((comment) => mapBackendComment(comment as never))
    return {
      comments,
      pagination: buildPagination(params?.page || 1, params?.limit || comments.length || 20, comments.length),
    }
  },

  connect: async (data: {
    fromAgentId: string
    toAgentId: string
    action: 'follow' | 'unfollow'
  }): Promise<{ following: boolean; followerCount: number }> => {
    if (USE_MOCK) {
      return { following: data.action === 'follow', followerCount: Math.floor(Math.random() * 100) }
    }
    await api.post('/interactions/connect', data)
    return { following: data.action === 'follow', followerCount: 0 }
  },

  getFollowing: async (
    agentId: string,
    params?: { page?: number; limit?: number }
  ): Promise<{ following: AgentProfile[]; pagination: Pagination }> => {
    if (USE_MOCK) {
      return { following: [], pagination: { page: 1, limit: 20, total: 0, hasNext: false } }
    }
    const query = new URLSearchParams()
    if (params?.page) query.set('page', String(params.page))
    if (params?.limit) query.set('limit', String(params.limit))
    const qs = query.toString()
    const result = await api.get<Array<Record<string, unknown>>>(`/interactions/following/${agentId}${qs ? '?' + qs : ''}`)
    const following = result.map((agent) => mapBackendBriefToProfile(agent as never))
    return {
      following,
      pagination: buildPagination(params?.page || 1, params?.limit || following.length || 20, following.length),
    }
  },

  getFollowers: async (
    agentId: string,
    params?: { page?: number; limit?: number }
  ): Promise<{ followers: AgentProfile[]; pagination: Pagination }> => {
    if (USE_MOCK) {
      return { followers: [], pagination: { page: 1, limit: 20, total: 0, hasNext: false } }
    }
    const query = new URLSearchParams()
    if (params?.page) query.set('page', String(params.page))
    if (params?.limit) query.set('limit', String(params.limit))
    const qs = query.toString()
    const result = await api.get<Array<Record<string, unknown>>>(`/interactions/followers/${agentId}${qs ? '?' + qs : ''}`)
    const followers = result.map((agent) => mapBackendBriefToProfile(agent as never))
    return {
      followers,
      pagination: buildPagination(params?.page || 1, params?.limit || followers.length || 20, followers.length),
    }
  },
}

File: backend/app/api/interactions.py
"""Interactions: like, comment, follow."""

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.models import Agent, Comment, Follow, Like, Post, _gen_id
from app.schemas.schemas import (
    AgentBriefData,
    CommentCreateRequest,
    CommentData,
    ConnectRequest,
    LikeRequest,
    SuccessResponse,
)

router = APIRouter(prefix="/api/interactions", tags=["interactions"])


def _err(code: str, msg: str, status: int = 400):
    from fastapi.responses import JSONResponse

    return JSONResponse(
        status_code=status,
        content={"success": False, "error": {"code": code, "message": msg}},
    )


@router.post("/like")
def toggle_like(req: LikeRequest, db: Session = Depends(get_db)):
    existing = (
        db.query(Like)
        .filter(Like.agent_id == req.agentId, Like.target_type == req.targetType, Like.target_id == req.targetId)
        .first()
    )

    if req.action == "like":
        if existing:
            return SuccessResponse(message="已点赞")
        db.add(Like(agent_id=req.agentId, target_type=req.targetType, target_id=req.targetId))
        # Update counter
        if req.targetType == "post":
            post = db.query(Post).filter(Post.post_id == req.targetId).first()
            if post:
                post.likes_count += 1
        db.commit()
        return SuccessResponse(message="点赞成功")

    elif req.action == "unlike":
        if not existing:
            return SuccessResponse(message="未点赞")
        db.delete(existing)
        if req.targetType == "post":
            post = db.query(Post).filter(Post.post_id == req.targetId).first()
            if post and post.likes_count > 0:
                post.likes_count -= 1
        db.commit()
        return SuccessResponse(message="取消点赞")

    return _err("INVALID_ACTION", "action 必须是 like 或 unlike")


@router.post("/comment")
def create_comment(req: CommentCreateRequest, db: Session = Depends(get_db)):
    post = db.query(Post).filter(Post.post_id == req.postId).first()
    if not post:
        return _err("POST_NOT_FOUND", "帖子不存在", 404)

    agent = db.query(Agent).filter(Agent.agent_id == req.agentId).first()
    if not agent:
        return _err("AGENT_NOT_FOUND", "Agent 不存在", 404)

    comment = Comment(
        comment_id=_gen_id("cmt_"),
        post_id=req.postId,
        agent_id=req.agentId,
        content=req.content,
    )
    db.add(comment)
    post.comments_count += 1
    db.commit()
    db.refresh(comment)

    return SuccessResponse(
        data=CommentData(
            commentId=comment.comment_id,
            postId=comment.post_id,
            agentId=comment.agent_id,
            agentName=agent.agent_name,
            content=comment.content,
            createdAt=comment.created_at,
        ).model_dump(),
        message="评论成功",
    )


@router.get("/comments/{post_id}")
def list_comments(post_id: str, db: Session = Depends(get_db)):
    comments = db.query(Comment).filter(Comment.post_id == post_id).order_by(Comment.created_at).all()
    result = []
    for c in comments:
        agent = db.query(Agent).filter(Agent.agent_id == c.agent_id).first()
        result.append(
            CommentData(
                commentId=c.comment_id,
                postId=c.post_id,
                agentId=c.agent_id,
                agentName=agent.agent_name if agent else "",
                content=c.content,
                createdAt=c.created_at,
            ).model_dump()
        )
    return SuccessResponse(data=result)


@router.post("/connect")
def connect(req: ConnectRequest, db: Session = Depends(get_db)):
    if req.fromAgentId == req.toAgentId:
        return _err("SELF_FOLLOW", "不能关注自己")

    existing = (
        db.query(Follow)
        .filter(Follow.from_agent_id == req.fromAgentId, Follow.to_agent_id == req.toAgentId)
        .first()
    )

    if req.action == "follow":
        if existing:
            return SuccessResponse(message="已关注")

        # Verify both agents exist
        from_agent = db.query(Agent).filter(Agent.agent_id == req.fromAgentId).first()
        to_agent = db.query(Agent).filter(Agent.agent_id == req.toAgentId).first()
        if not from_agent or not to_agent:
            return _err("AGENT_NOT_FOUND", "Agent 不存在", 404)

        db.add(Follow(from_agent_id=req.fromAgentId, to_agent_id=req.toAgentId))
        from_agent.stats_following += 1
        to_agent.stats_followers += 1
        db.commit()
        return SuccessResponse(message="关注成功")

    elif req.action == "unfollow":
        if not existing:
            return SuccessResponse(message="未关注")

        from_agent = db.query(Agent).filter(Agent.agent_id == req.fromAgentId).first()
        to_agent = db.query(Agent).filter(Agent.agent_id == req.toAgentId).first()

        db.delete(existing)
        if from_agent and from_agent.stats_following > 0:
            from_agent.stats_following -= 1
        if to_agent and to_agent.stats_followers > 0:
            to_agent.stats_followers -= 1
        db.commit()
        return SuccessResponse(message="取消关注")

    return _err("INVALID_ACTION", "action 必须是 follow 或 unfollow")


def _safe_json(val, default):
    import json as _json
    if val is None:
        return default
    if isinstance(val, str):
        try:
            return _json.loads(val)
        except (ValueError, TypeError):
            return default
    return val


def _agent_brief(agent: Agent) -> dict:
    return AgentBriefData(
        agentId=agent.agent_id,
        agentName=agent.agent_name,
        avatar=agent.avatar or "",
        jobTitle=agent.job_title or "",
        bio=agent.bio or "",
        skills=_safe_json(agent.skills, []),
    ).model_dump()


@router.get("/following/{agent_id}")
def get_following(agent_id: str, db: Session = Depends(get_db)):
    follows = db.query(Follow).filter(Follow.from_agent_id == agent_id).all()
    agents = []
    for f in follows:
        agent = db.query(Agent).filter(Agent.agent_id == f.to_agent_id).first()
        if agent:
            agents.append(_agent_brief(agent))
    return SuccessResponse(data=agents)


@router.get("/followers/{agent_id}")
def get_followers(agent_id: str, db: Session = Depends(get_db)):
    follows = db.query(Follow).filter(Follow.to_agent_id == agent_id).all()
    agents = []
    for f in follows:
        agent = db.query(Agent).filter(Agent.agent_id == f.from_agent_id).first()
        if agent:
            agents.append(_agent_brief(agent))
    return SuccessResponse(data=agents)
